File: 2020/fsharp/AOC/Day10.fs
module AOC.Day10

let diff (arr: int []) = 
    Array.map2 (-) arr.[1..] arr.[..((Array.length arr) - 2)]

let countDigit digit arr = 
    arr |> Seq.filter ((=) digit) |> Seq.length

let part1 data =
    let differences = (Array.concat [[|0|]; data]) |> Array.sort |> diff 
    (countDigit 1 differences) * ((countDigit 3 differences) + 1)

let part2 data = 
    let differences = (Array.concat [[|0|]; data]) |> Array.sort |> diff 
    let groupLens = Array.fold (fun (groups: int []) d -> if d = 1 then 
                                                            let idx = (Array.length groups) - 1
                                                            groups.[idx] <- groups.[idx] + 1
                                                            groups
                                                          else Array.concat [groups; [|0|]])
                                [|0|]
                                differences
    let num2s = countDigit 2 groupLens
    let num3s = countDigit 3 groupLens
    let num4s = countDigit 4 groupLens
    (bigint 2)**num2s * (bigint 4)**num3s * (bigint 7)**num4s

File: 2020/julia/AOC/src/day10.jl
module Day10
import AOC.Parsers

function part1(;data=Parsers.day10())
	differences = diff(sort(vcat([0], data)))
	count(differences .== 1) * (count(differences .== 3) + 1)
end

function part2(;data=Parsers.day10())
	differences = diff(sort(vcat([0], data)))
	grouplens = [0]
	for d in differences 
		if d == 1
			grouplens[end] += 1
		else
			push!(grouplens, 0)
		end
	end
	num2s = count(grouplens .== 2)
	num3s = count(grouplens .== 3)
	num4s = count(grouplens .== 4)
	2^num2s * 4^num3s * 7^num4s
end

end